File: R/08_coexpression_networks.R
# 08_coexpression_networks.R
# ============================
# Gene co-expression network analysis using WGCNA.
#
# WHAT THIS SCRIPT DOES:
#   1. Subset to target cell type and build metacells (k-NN aggregation)
#      — metacells average k nearest neighbours to reduce technical zeros
#   2. Soft-thresholding power selection (scale-free topology test)
#   3. WGCNA network construction and module detection
#   4. Hub gene identification per module (by intramodular connectivity kME)
#   5. Pathway enrichment per module via enrichR
#   6. Save all results
#
# WHY METACELLS?
#   Single-cell data is 80-90% sparse. Averaging k nearest neighbours
#   in PCA space fills technical zeros without losing cell-type specificity.
#
# HOW TO USE:
#   1. Edit CONFIGURATION below.
#   2. Run:  Rscript 08_coexpression_networks.R
#
# OUTPUT:
#   results/08_coexpression/module_assignments.csv
#   results/08_coexpression/module_eigengenes.csv
#   results/08_coexpression/hub_genes.csv
#   results/08_coexpression/module_enrichment.csv
#   results/08_coexpression/plots/

suppressMessages({
  library(Seurat)
  library(WGCNA)
  library(FNN)
  library(ggplot2)
  library(dplyr)
  library(enrichR)
})

options(stringsAsFactors = FALSE)
WGCNA::allowWGCNAThreads()


# ==============================================================================
#  CONFIGURATION — Edit this section before running
# ==============================================================================

INPUT_PATH    <- "results/03_annotation/seurat_annotated.rds"
CELL_TYPE_COL <- "cell_type"
SAMPLE_COL    <- "sample"
CONDITION_COL <- "condition"    # used for module-trait correlation

# --- Target cell type ---------------------------------------------------------
# Leave "" to use all cells (not recommended — very slow and noisy).
TARGET_CELL_TYPE <- "CD4+ T cell"

# --- Metacell parameters ------------------------------------------------------
METACELL_K        <- 20L     # neighbours to average per metacell
MAX_SHARED        <- 10L     # max cell overlap between two metacells
MIN_EXPR_FRACTION <- 0.05    # keep genes expressed in >= 5% of metacells

# --- WGCNA parameters ---------------------------------------------------------
# Soft-thresholding power β.
# Set to 0 to auto-detect (chooses lowest β where scale-free R² >= 0.85).
SOFT_POWER     <- 0L
MIN_MODULE_SIZE <- 30L
MERGE_CUT_HEIGHT <- 0.25   # modules with correlation > (1 - 0.25) are merged

# --- Pathway enrichment -------------------------------------------------------
RUN_ENRICHR        <- TRUE
MODULE_TOP_N_GENES <- 50L    # top genes per module submitted to enrichR
ENRICHR_LIBS       <- c("MSigDB_Hallmark_2020", "KEGG_2021_Human")

# --- Output -------------------------------------------------------------------
OUTPUT_DIR <- "results/08_coexpression"


# ==============================================================================
#  HELPERS
# ==============================================================================

header <- function(title) {
  message()
  message(strrep("=", 60))
  message("  ", title)
  message(strrep("=", 60))
}


# ==============================================================================
#  STEP 1 — LOAD & SUBSET
# ==============================================================================

load_and_subset <- function() {
  header("Step 1 — Load & Subset Data")
  if (!file.exists(INPUT_PATH)) stop("File not found: ", INPUT_PATH)
  seu <- readRDS(INPUT_PATH)
  message(sprintf("  Loaded %d cells x %d genes", ncol(seu), nrow(seu)))

  if (nchar(TARGET_CELL_TYPE) > 0L) {
    mask <- seu@meta.data[[CELL_TYPE_COL]] == TARGET_CELL_TYPE
    seu  <- seu[, mask]
    if (ncol(seu) < 100L) {
      stop(sprintf("Only %d cells for '%s'. Need >= 100 cells.",
                   ncol(seu), TARGET_CELL_TYPE))
    }
    message(sprintf("  Subset to '%s': %d cells", TARGET_CELL_TYPE, ncol(seu)))
  } else {
    message("  Using all cells (TARGET_CELL_TYPE is empty)")
  }
  seu
}


# ==============================================================================
#  STEP 2 — METACELL CONSTRUCTION
# ==============================================================================

build_metacells <- function(seu) {
  header("Step 2 — Metacell Construction")
  message(sprintf("  k = %d nearest neighbours per metacell", METACELL_K))

  # Use raw counts
  counts_mat <- as.matrix(GetAssayData(seu, layer = "counts"))   # genes x cells
  counts_mat <- t(counts_mat)                                     # cells x genes

  # PCA embedding for kNN
  if ("pca" %in% names(seu@reductions)) {
    pca_mat <- Embeddings(seu, "pca")
  } else {
    pca_mat <- counts_mat[, seq_len(min(50L, ncol(counts_mat)))]
  }

  samples <- if (SAMPLE_COL %in% colnames(seu@meta.data)) {
    unique(as.character(seu@meta.data[[SAMPLE_COL]]))
  } else {
    "all"
  }

  mc_rows <- list()
  mc_meta <- list()

  for (s in samples) {
    if (s == "all") {
      idx <- seq_len(nrow(pca_mat))
    } else {
      idx <- which(seu@meta.data[[SAMPLE_COL]] == s)
    }

    pca_s    <- pca_mat[idx, , drop = FALSE]
    counts_s <- counts_mat[idx, , drop = FALSE]

    # kNN
    nn <- FNN::get.knn(pca_s, k = METACELL_K)$nn.index

    used      <- logical(nrow(pca_s))
    n_target  <- max(1L, nrow(pca_s) %/% METACELL_K)

    for (i in seq_len(nrow(pca_s))) {
      knn_idx <- nn[i, ]
      overlap <- sum(used[knn_idx])
      if (overlap > MAX_SHARED) next

      mc_rows <- c(mc_rows, list(colMeans(counts_s[knn_idx, , drop = FALSE])))
      mc_meta <- c(mc_meta, list(data.frame(sample = s)))
      used[knn_idx] <- TRUE
      if (length(mc_rows) >= n_target) break
    }
  }

  if (length(mc_rows) == 0L) stop("Metacell construction failed. Try reducing METACELL_K.")

  mc_matrix <- do.call(rbind, mc_rows)   # metacells x genes
  mc_df     <- data.frame(do.call(rbind, mc_meta))

  # Gene filter: expressed in >= MIN_EXPR_FRACTION of metacells
  expr_frac <- colMeans(mc_matrix > 0)
  keep      <- expr_frac >= MIN_EXPR_FRACTION
  mc_matrix <- mc_matrix[, keep, drop = FALSE]

  message(sprintf("  Built %d metacells x %d genes", nrow(mc_matrix), ncol(mc_matrix)))
  list(matrix = mc_matrix, meta = mc_df)
}


# ==============================================================================
#  STEP 3 — SOFT-THRESHOLDING POWER SELECTION
# ==============================================================================

select_soft_power <- function(mc_matrix, plot_dir) {
  header("Step 3 — Soft-Thresholding Power Selection")

  if (SOFT_POWER > 0L) {
    message(sprintf("  Using manually set SOFT_POWER = %d", SOFT_POWER))
    return(SOFT_POWER)
  }

  powers  <- 1:30
  sft     <- pickSoftThreshold(mc_matrix, powerVector = powers,
                                networkType = "unsigned", verbose = 0)

  # Plot
  df <- data.frame(Power = sft$fitIndices$Power,
                   R2    = -sign(sft$fitIndices$slope) * sft$fitIndices$SFT.R.sq,
                   MeanK = sft$fitIndices$mean.k.)

  p1 <- ggplot(df, aes(Power, R2)) +
    geom_line() + geom_point() +
    geom_hline(yintercept = 0.85, linetype = "dashed", colour = "red") +
    labs(title = "Scale-Free Fit (R²)", x = "Soft Power (β)", y = "R²") +
    theme_classic()

  p2 <- ggplot(df, aes(Power, MeanK)) +
    geom_line() + geom_point() +
    labs(title = "Mean Connectivity", x = "Soft Power (β)", y = "Mean k") +
    theme_classic()

  ggsave(file.path(plot_dir, "00_soft_power_selection.pdf"),
         p1 + p2, width = 10, height = 4)

  auto_power <- sft$powerEstimate
  if (is.na(auto_power)) {
    auto_power <- 6L
    message("  R² >= 0.85 not reached — defaulting to β = 6. Consider reviewing the plot.")
  }
  message(sprintf("  Auto-selected soft power β = %d", auto_power))
  as.integer(auto_power)
}


# ==============================================================================
#  STEP 4 — WGCNA MODULE DETECTION
# ==============================================================================

run_wgcna <- function(mc_matrix, selected_power) {
  header("Step 4 — WGCNA Network Construction & Module Detection")

  # blockwiseModules handles large gene sets by processing in blocks
  net <- blockwiseModules(
    datExpr          = mc_matrix,
    power            = selected_power,
    networkType      = "unsigned",
    TOMType          = "unsigned",
    minModuleSize    = MIN_MODULE_SIZE,
    mergeCutHeight   = MERGE_CUT_HEIGHT,
    reassignThreshold = 0,
    numericLabels    = FALSE,
    pamRespectsDendro = FALSE,
    saveTOMs         = FALSE,
    verbose          = 0
  )

  n_modules <- length(unique(net$colors)) - 1L  # exclude grey (unassigned)
  message(sprintf("  %d co-expression modules detected", n_modules))
  net
}


# ==============================================================================
#  STEP 5 — HUB GENES (kME)
# ==============================================================================

get_hub_genes <- function(mc_matrix, net) {
  header("Step 5 — Hub Genes (Intramodular Connectivity kME)")

  kme_mat <- signedKME(mc_matrix, net$MEs, outputColumnName = "kME")

  modules <- unique(net$colors[net$colors != "grey"])
  rows    <- list()

  for (mod in modules) {
    col     <- paste0("kME", mod)
    if (!col %in% colnames(kme_mat)) next
    gene_mask <- net$colors == mod
    mod_kme   <- kme_mat[gene_mask, col, drop = TRUE]
    top_idx   <- order(mod_kme, decreasing = TRUE)[seq_len(min(10L, sum(gene_mask)))]
    rows[[mod]] <- data.frame(
      gene   = rownames(kme_mat)[gene_mask][top_idx],
      module = mod,
      kME    = mod_kme[top_idx]
    )
  }

  hub_df <- bind_rows(rows)
  message("  Top 3 hub genes per module:")
  for (mod in modules) {
    top3 <- head(hub_df$gene[hub_df$module == mod], 3L)
    message(sprintf("    %s: %s", mod, paste(top3, collapse = ", ")))
  }
  hub_df
}


# ==============================================================================
#  STEP 6 — MODULE EIGENGENE PLOT
# ==============================================================================

save_eigengene_plot <- function(net, plot_dir) {
  me_corr <- cor(net$MEs)

  me_df <- as.data.frame(as.table(me_corr))
  colnames(me_df) <- c("Module1", "Module2", "Correlation")

  p <- ggplot(me_df, aes(Module1, Module2, fill = Correlation)) +
    geom_tile() +
    scale_fill_gradient2(low = "#4C72B0", mid = "white", high = "#C44E52",
                         midpoint = 0, limits = c(-1, 1)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 7),
          axis.text.y = element_text(size = 7)) +
    labs(title = "Module Eigengene Correlation Matrix",
         x = NULL, y = NULL)

  ggsave(file.path(plot_dir, "01_eigengene_correlation.pdf"), p, width = 8, height = 7)
}


# ==============================================================================
#  STEP 7 — PATHWAY ENRICHMENT PER MODULE
# ==============================================================================

run_module_enrichment <- function(mc_matrix, net) {
  header("Step 7 — Pathway Enrichment per Module (enrichR)")

  if (!RUN_ENRICHR) {
    message("  Skipped (RUN_ENRICHR = FALSE)")
    return(NULL)
  }

  modules <- unique(net$colors[net$colors != "grey"])
  results <- list()

  for (mod in modules) {
    gene_mask <- net$colors == mod
    genes     <- colnames(mc_matrix)[gene_mask]
    genes     <- head(genes, MODULE_TOP_N_GENES)
    if (length(genes) < 5L) next

    tryCatch({
      enr <- enrichr(genes, ENRICHR_LIBS)
      for (lib in names(enr)) {
        r          <- head(enr[[lib]], 5L)
        r$module   <- mod
        r$n_genes  <- length(genes)
        r$library  <- lib
        results[[paste(mod, lib)]] <- r
      }
      message(sprintf("  Module %s: enrichment done", mod))
    }, error = function(e) {
      message(sprintf("  Module %s: enrichment failed — %s", mod, conditionMessage(e)))
    })
  }

  if (length(results) == 0L) return(NULL)
  bind_rows(results)
}


# ==============================================================================
#  MAIN
# ==============================================================================

main <- function() {
  dir.create(OUTPUT_DIR, recursive = TRUE, showWarnings = FALSE)
  plot_dir <- file.path(OUTPUT_DIR, "plots")
  dir.create(plot_dir, recursive = TRUE, showWarnings = FALSE)

  message()
  message(strrep("=", 60))
  message("  scRNA-seq Co-expression Networks (WGCNA)")
  message(strrep("=", 60))
  message(sprintf("  Input      : %s", INPUT_PATH))
  message(sprintf("  Cell type  : %s", if (nchar(TARGET_CELL_TYPE) > 0L) TARGET_CELL_TYPE else "all cells"))
  message(sprintf("  Metacell k : %d", METACELL_K))
  message(sprintf("  Output     : %s", OUTPUT_DIR))

  seu            <- load_and_subset()
  mc             <- build_metacells(seu)
  power          <- select_soft_power(mc$matrix, plot_dir)
  net            <- run_wgcna(mc$matrix, power)
  hub_genes      <- get_hub_genes(mc$matrix, net)
  save_eigengene_plot(net, plot_dir)
  enrichment_df  <- run_module_enrichment(mc$matrix, net)

  header("Save")

  # Module assignments
  assign_df <- data.frame(gene = colnames(mc$matrix), module = net$colors)
  write.csv(assign_df,  file.path(OUTPUT_DIR, "module_assignments.csv"),  row.names = FALSE)
  message("  Module assignments -> module_assignments.csv")

  # Module eigengenes
  write.csv(net$MEs,    file.path(OUTPUT_DIR, "module_eigengenes.csv"))
  message("  Module eigengenes  -> module_eigengenes.csv")

  # Hub genes
  write.csv(hub_genes,  file.path(OUTPUT_DIR, "hub_genes.csv"),           row.names = FALSE)
  message("  Hub genes          -> hub_genes.csv")

  # Pathway enrichment
  if (!is.null(enrichment_df)) {
    write.csv(enrichment_df, file.path(OUTPUT_DIR, "module_enrichment.csv"), row.names = FALSE)
    message("  Module enrichment  -> module_enrichment.csv")
  }

  message()
  message("  Pipeline complete. All results in: ", OUTPUT_DIR)
  message()
}

main()

File: Python/08_coexpression_networks.py
#!/usr/bin/env python3
"""
08_coexpression_networks.py
============================
Gene co-expression network analysis using PyWGCNA.

WHAT THIS SCRIPT DOES:
  1. Subset to target cell type and build metacells (k-NN aggregation)
     — metacells average k nearest neighbours to fill technical zeros
  2. Soft-thresholding power selection (scale-free topology test)
  3. WGCNA network construction and module detection
  4. Hub gene identification per module (by intramodular connectivity)
  5. Pathway enrichment per module via Enrichr
  6. Save all results

WHY METACELLS?
  Single-cell data is 80-90% sparse (zeros). Averaging k nearest neighbours
  in PCA space fills in technical zeros without losing cell-type specificity.

HOW TO USE:
  1. Edit CONFIGURATION below.
  2. Run:  python 08_coexpression_networks.py

OUTPUT:
  results/08_coexpression/module_assignments.csv
  results/08_coexpression/module_eigengenes.csv
  results/08_coexpression/hub_genes.csv
  results/08_coexpression/module_enrichment.csv
  results/08_coexpression/plots/
"""

import sys
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import scanpy as sc
import plotly.express as px

sys.path.insert(0, str(Path("..") / "src"))
from sc_workflow.utils import setup_scanpy, save_adata

warnings.filterwarnings("ignore")
setup_scanpy()


# ==============================================================================
#  CONFIGURATION — Edit this section before running
# ==============================================================================

INPUT_PATH = "results/03_annotation/adata_annotated.h5ad"
CELL_TYPE_COL = "cell_type"
SAMPLE_COL = "sample"
CONDITION_COL = "condition"   # used for module-trait correlation

# --- Target cell type ---------------------------------------------------------
# Which cell type to build the network for.
# Leave empty ("") to use all cells (not recommended — very slow and noisy).
TARGET_CELL_TYPE = "CD4+ T cell"

# --- Metacell parameters ------------------------------------------------------
# k = number of nearest neighbours to average per metacell.
# Rule of thumb: k = 20-30 for most datasets.
METACELL_K = 20

# Two metacells sharing more than this many cells are not both kept.
MAX_SHARED = 10

# Only keep genes expressed (count > 0) in at least this fraction of metacells.
MIN_EXPR_FRACTION = 0.05

# --- WGCNA parameters --------------------------------------------------------
# Soft-thresholding power β.
# Set to 0 to auto-detect (chooses lowest β where scale-free R² ≥ 0.85).
SOFT_POWER = 0

# Minimum number of genes per module.
MIN_MODULE_SIZE = 30

# Modules with eigengene correlation > (1 - merge_cut_height) are merged.
MERGE_CUT_HEIGHT = 0.25

# --- Pathway enrichment per module --------------------------------------------
RUN_MODULE_ENRICHMENT = True
MODULE_TOP_N_GENES = 50   # top genes per module to submit to Enrichr

# --- Output ------------------------------------------------------------------
OUTPUT_DIR = "results/08_coexpression"


# ==============================================================================
#  HELPERS
# ==============================================================================

def print_header(title):
    print()
    print("=" * 60)
    print(f"  {title}")
    print("=" * 60)


def print_ok(msg):
    print(f"  [OK]  {msg}")


def print_info(msg):
    print(f"  [--]  {msg}")


# ==============================================================================
#  STEP 1 — LOAD & SUBSET
# ==============================================================================

def load_and_subset():
    print_header("Step 1 — Load & Subset Data")
    p = Path(INPUT_PATH)
    if not p.exists():
        print(f"  [!!]  File not found: {p}")
        sys.exit(1)
    adata = sc.read_h5ad(p)
    print_ok(f"Loaded {adata.n_obs:,} cells × {adata.n_vars:,} genes")

    if TARGET_CELL_TYPE:
        mask = adata.obs[CELL_TYPE_COL] == TARGET_CELL_TYPE
        adata_sub = adata[mask].copy()
        if adata_sub.n_obs < 100:
            print(f"  [!!]  Only {adata_sub.n_obs} cells for '{TARGET_CELL_TYPE}'.")
            print("        Need >= 100 cells (>=200 recommended).")
            sys.exit(1)
        print_ok(f"Subset to '{TARGET_CELL_TYPE}': {adata_sub.n_obs:,} cells")
    else:
        adata_sub = adata.copy()
        print_info("Using all cells (TARGET_CELL_TYPE is empty)")

    return adata_sub


# ==============================================================================
#  STEP 2 — METACELL CONSTRUCTION
# ==============================================================================

def build_metacells(adata_sub):
    print_header("Step 2 — Metacell Construction")
    print_info(f"k={METACELL_K} nearest neighbours per metacell")

    from sklearn.neighbors import NearestNeighbors
    import scipy.sparse as sp

    # Use raw counts if available
    X = adata_sub.layers["counts"] if "counts" in adata_sub.layers else adata_sub.X
    if sp.issparse(X):
        X = X.toarray()
    X = X.astype(float)

    samples = (adata_sub.obs[SAMPLE_COL].unique().tolist()
               if SAMPLE_COL in adata_sub.obs.columns else ["all"])

    metacell_mats = []
    metacell_meta = []

    for s in samples:
        if s == "all":
            s_mask = np.ones(adata_sub.n_obs, dtype=bool)
        else:
            s_mask = (adata_sub.obs[SAMPLE_COL] == s).values

        X_s = X[s_mask]
        pca_s = (adata_sub.obsm["X_pca"][s_mask]
                 if "X_pca" in adata_sub.obsm else X_s[:, :50])

        n_mc_target = max(1, X_s.shape[0] // METACELL_K)
        nbrs = NearestNeighbors(n_neighbors=METACELL_K, algorithm="ball_tree").fit(pca_s)
        _, indices = nbrs.kneighbors(pca_s)

        used = set()
        for i in range(X_s.shape[0]):
            knn = indices[i]
            overlap = sum(j in used for j in knn)
            if overlap > MAX_SHARED:
                continue
            metacell_mats.append(X_s[knn].mean(axis=0))
            metacell_meta.append({"sample": s})
            used.update(knn.tolist())
            if len(metacell_mats) >= n_mc_target:
                break

    if not metacell_mats:
        print("  [!!]  Metacell construction failed. Try reducing METACELL_K.")
        sys.exit(1)

    mc_matrix = np.vstack(metacell_mats)   # metacells × genes
    mc_meta = pd.DataFrame(metacell_meta)

    # Gene filter
    expr_frac = (mc_matrix > 0).mean(axis=0)
    gene_mask = expr_frac >= MIN_EXPR_FRACTION
    mc_matrix = mc_matrix[:, gene_mask]
    genes_kept = adata_sub.var_names[gene_mask].tolist()

    metacell_df = pd.DataFrame(mc_matrix, columns=genes_kept)
    print_ok(f"Built {metacell_df.shape[0]:,} metacells × {metacell_df.shape[1]:,} genes")
    print_ok(f"(Genes expressed in >= {MIN_EXPR_FRACTION*100:.0f}% of metacells retained)")
    return metacell_df, mc_meta, genes_kept


# ==============================================================================
#  STEP 3 — SOFT-THRESHOLDING POWER SELECTION
# ==============================================================================

def select_soft_power(metacell_df):
    print_header("Step 3 — Soft-Thresholding Power Selection")

    if SOFT_POWER > 0:
        print_ok(f"Using manually set SOFT_POWER = {SOFT_POWER}")
        return SOFT_POWER

    powers = list(range(1, 31))
    r2_scores = []

    corr = np.corrcoef(metacell_df.values.T)   # genes × genes
    np.fill_diagonal(corr, 0)
    corr_abs = np.abs(corr)

    for beta in powers:
        adj = corr_abs ** beta
        k = adj.sum(axis=1) - 1   # connectivity degree
        log_k = np.log10(k + 1e-10)
        hist_counts, bin_edges = np.histogram(k, bins=20)
        log_pk = np.log10(hist_counts / (len(k) + 1e-10) + 1e-10)
        x = np.linspace(log_k.min(), log_k.max(), 20)
        valid = np.isfinite(log_pk) & (log_pk > -np.inf)
        if valid.sum() > 3:
            fit = np.polyfit(x[valid], log_pk[valid], 1)
            yhat = np.polyval(fit, x[valid])
            ss_res = ((log_pk[valid] - yhat) ** 2).sum()
            ss_tot = ((log_pk[valid] - log_pk[valid].mean()) ** 2).sum()
            r2 = max(0.0, 1 - ss_res / (ss_tot + 1e-10))
        else:
            r2 = 0.0
        r2_scores.append(r2)

    sft_df = pd.DataFrame({"Power": powers, "R2": r2_scores})
    auto_candidates = sft_df[sft_df["R2"] >= 0.85]
    auto_power = int(auto_candidates["Power"].iloc[0]) if len(auto_candidates) > 0 else 6

    print()
    print("  Power  R²")
    print("  ─────  ────")
    for _, row in sft_df.iterrows():
        marker = " <-- selected" if row["Power"] == auto_power else ""
        print(f"  {int(row['Power']):>5}  {row['R2']:.3f}{marker}")

    print_ok(f"Auto-selected soft power β = {auto_power} "
             f"(R² = {sft_df.loc[sft_df['Power']==auto_power, 'R2'].values[0]:.3f})")
    return auto_power


# ==============================================================================
#  STEP 4 — WGCNA MODULE DETECTION
# ==============================================================================

def run_wgcna(metacell_df, mc_meta, selected_power):
    print_header("Step 4 — WGCNA Network Construction & Module Detection")

    try:
        import PyWGCNA as pw
    except ImportError:
        print("  [!!]  PyWGCNA not installed. Run: pip install PyWGCNA")
        sys.exit(1)

    name = (TARGET_CELL_TYPE.replace(" ", "_").replace("+", "pos")
            if TARGET_CELL_TYPE else "all_cells")

    wgcna = pw.WGCNA(
        name=name,
        species="human",
        geneExp=metacell_df.T,   # PyWGCNA expects genes × samples
        TPMcutoff=0,
        save=False,
    )

    wgcna.preprocess()
    wgcna.findModules(
        power=selected_power,
        minModuleSize=MIN_MODULE_SIZE,
        deepSplit=4,
        mergeCutHeight=MERGE_CUT_HEIGHT,
    )
    wgcna.sampleInfo = mc_meta.copy()

    n_modules = len(wgcna.moduleColors.unique()) - 1  # exclude grey (unassigned)
    print_ok(f"WGCNA complete: {n_modules} co-expression modules detected "
             f"across {metacell_df.shape[1]:,} genes")
    return wgcna


# ==============================================================================
#  STEP 5 — HUB GENES & VISUALISATION
# ==============================================================================

def get_hub_genes(wgcna):
    print_header("Step 5 — Hub Genes (Intramodular Connectivity)")

    kme = wgcna.kMEtable
    hub_genes = (
        kme.reset_index()
        .melt(id_vars="index", var_name="module", value_name="kME")
        .rename(columns={"index": "gene"})
        .sort_values("kME", ascending=False)
        .groupby("module")
        .head(10)
        .sort_values(["module", "kME"], ascending=[True, False])
    )

    print()
    print("  Top 3 hub genes per module:")
    for module, group in hub_genes.groupby("module"):
        top3 = group.head(3)["gene"].tolist()
        print(f"    {module}: {', '.join(top3)}")

    return hub_genes


def save_eigengene_plot(wgcna, plot_dir):
    plot_dir = Path(plot_dir)
    plot_dir.mkdir(parents=True, exist_ok=True)

    me = wgcna.MEs
    fig = px.imshow(
        me.corr(),
        color_continuous_scale="RdBu_r",
        zmin=-1, zmax=1,
        title="Module Eigengene Correlation Matrix",
        template="plotly_white",
    )
    fig.update_layout(height=500)
    fig.write_html(plot_dir / "01_eigengene_correlation.html")
    print_ok(f"Eigengene plot saved → {plot_dir}/01_eigengene_correlation.html")


# ==============================================================================
#  STEP 6 — PATHWAY ENRICHMENT PER MODULE
# ==============================================================================

def run_module_enrichment(wgcna):
    print_header("Step 6 — Pathway Enrichment per Module (Enrichr)")

    if not RUN_MODULE_ENRICHMENT:
        print_info("Module enrichment skipped (RUN_MODULE_ENRICHMENT = False)")
        return None

    try:
        import gseapy as gp
    except ImportError:
        print("  [!!]  gseapy not installed. Run: pip install gseapy")
        return None

    results = []
    modules = [m for m in wgcna.moduleColors.unique() if m != "grey"]

    for mod in modules:
        genes = wgcna.datExpr.columns[wgcna.moduleColors == mod].tolist()
        genes = genes[:MODULE_TOP_N_GENES]
        if len(genes) < 5:
            continue
        try:
            enr = gp.enrichr(
                gene_list=genes,
                gene_sets=["MSigDB_Hallmark_2020", "KEGG_2021_Human"],
                organism="human",
                outdir=None,
                verbose=False,
            )
            r = enr.results.head(5).copy()
            r["module"] = mod
            r["n_genes"] = len(genes)
            results.append(r)
            print_ok(f"Module {mod}: {len(r)} enrichment terms")
        except Exception as e:
            print_info(f"Module {mod}: enrichment failed — {e}")

    if not results:
        print_info("No enrichment results returned")
        return None

    return pd.concat(results)


# ==============================================================================
#  MAIN
# ==============================================================================

def main():
    print()
    print("=" * 60)
    print("  scRNA-seq Co-expression Networks (PyWGCNA)")
    print("=" * 60)
    print(f"  Input       : {INPUT_PATH}")
    print(f"  Cell type   : {TARGET_CELL_TYPE or 'all cells'}")
    print(f"  Metacell k  : {METACELL_K}")
    print(f"  Output      : {OUTPUT_DIR}")

    out_dir = Path(OUTPUT_DIR)
    plot_dir = out_dir / "plots"
    out_dir.mkdir(parents=True, exist_ok=True)

    adata_sub = load_and_subset()
    metacell_df, mc_meta, genes_kept = build_metacells(adata_sub)
    selected_power = select_soft_power(metacell_df)
    wgcna = run_wgcna(metacell_df, mc_meta, selected_power)
    hub_genes = get_hub_genes(wgcna)
    save_eigengene_plot(wgcna, plot_dir)
    enrichment_df = run_module_enrichment(wgcna)

    print_header("Save")

    # Module assignments
    module_path = out_dir / "module_assignments.csv"
    pd.DataFrame({
        "gene": wgcna.datExpr.columns,
        "module": wgcna.moduleColors,
    }).to_csv(module_path, index=False)
    print_ok(f"Module assignments → {module_path}")

    # Module eigengenes
    me_path = out_dir / "module_eigengenes.csv"
    wgcna.MEs.to_csv(me_path)
    print_ok(f"Module eigengenes → {me_path}")

    # Hub genes
    hub_path = out_dir / "hub_genes.csv"
    hub_genes.to_csv(hub_path, index=False)
    print_ok(f"Hub genes → {hub_path}")

    # Pathway enrichment
    if enrichment_df is not None:
        enr_path = out_dir / "module_enrichment.csv"
        enrichment_df.to_csv(enr_path, index=False)
        print_ok(f"Module enrichment → {enr_path}")

    print()
    print("  Pipeline complete. All results saved to:", out_dir)
    print()


if __name__ == "__main__":
    main()
